File: web/packages/alloy-audio/src/dsp/lfo.ts
// Modulation LFO with delay + fade-in gate (vibrato that arrives late is
// the single most "played by a human" trick in the rompler book).
// Twin: Lfo.swift.

import { TWO_PI } from './dsp-types.js';

export type LfoShape = 'sine' | 'triangle';

export interface LfoParams {
  shape: LfoShape;
  rateHz: number;
  /** Seconds of silence before the LFO starts. */
  delay: number;
  /** Seconds to ramp depth 0 → 1 once started. */
  fadeIn: number;
}

export class Lfo {
  private phase = 0;
  private elapsed = 0;

  constructor(
    private readonly params: LfoParams,
    private readonly sampleRate: number,
  ) {}

  /** Next value in [−1, 1], gated by the delay/fade-in window. */
  nextSample(): number {
    const delaySamples = this.params.delay * this.sampleRate;
    const fadeSamples = this.params.fadeIn * this.sampleRate;
    const since = this.elapsed - delaySamples;
    this.elapsed += 1;
    if (since < 0) {
      return 0;
    }
    const gate = fadeSamples <= 0 ? 1 : Math.min(1, since / fadeSamples);
    const raw = this.params.shape === 'sine' ? Math.sin(TWO_PI * this.phase) : triangle(this.phase);
    this.phase += this.params.rateHz / this.sampleRate;
    this.phase -= Math.floor(this.phase);
    return raw * gate;
  }
}

/** Sine-aligned triangle: 0 → +1 → −1 → 0 across one cycle. */
function triangle(p: number): number {
  if (p < 0.25) {
    return 4 * p;
  }
  if (p < 0.75) {
    return 2 - 4 * p;
  }
  return 4 * p - 4;
}

File: swift/Sources/AlloyAudio/DSP/Lfo.swift
import Foundation

/// Modulation LFO with delay + fade-in gate. Twin of web src/dsp/lfo.ts.
public enum LfoShape {
    case sine
    case triangle
}

public struct LfoParams {
    public let shape: LfoShape
    public let rateHz: Double
    public let delay: Double
    public let fadeIn: Double

    public init(shape: LfoShape, rateHz: Double, delay: Double, fadeIn: Double) {
        self.shape = shape
        self.rateHz = rateHz
        self.delay = delay
        self.fadeIn = fadeIn
    }
}

public final class Lfo {
    private var phase = 0.0
    private var elapsed = 0.0
    private let params: LfoParams
    private let sampleRate: Double

    public init(params: LfoParams, sampleRate: Double) {
        self.params = params
        self.sampleRate = sampleRate
    }

    public func nextSample() -> Double {
        let delaySamples = params.delay * sampleRate
        let fadeSamples = params.fadeIn * sampleRate
        let since = elapsed - delaySamples
        elapsed += 1
        if since < 0 { return 0 }
        let gate = fadeSamples <= 0 ? 1 : min(1, since / fadeSamples)
        let raw: Double
        switch params.shape {
        case .sine:
            raw = sin(DspConstants.twoPi * phase)
        case .triangle:
            raw = Self.triangle(phase)
        }
        phase += params.rateHz / sampleRate
        phase -= phase.rounded(.down)
        return raw * gate
    }

    /// Sine-aligned triangle: 0 → +1 → −1 → 0 across one cycle.
    private static func triangle(_ p: Double) -> Double {
        if p < 0.25 { return 4 * p }
        if p < 0.75 { return 2 - 4 * p }
        return 4 * p - 4
    }
}
